File: backend/app/ai/gemini_scam_service.py
import os
import re
import json
from typing import Dict, Any, List
from backend.app.core.config import settings

class ThreatIndicators:
    def __init__(
        self,
        has_urgency_tactics: bool = False,
        has_suspicious_links: bool = False,
        has_impersonation: bool = False,
        requests_remote_access: bool = False,
        requests_credentials_or_otp: bool = False,
    ):
        self.has_urgency_tactics = has_urgency_tactics
        self.has_suspicious_links = has_suspicious_links
        self.has_impersonation = has_impersonation
        self.requests_remote_access = requests_remote_access
        self.requests_credentials_or_otp = requests_credentials_or_otp

    def to_dict(self) -> Dict[str, bool]:
        return {
            "hasUrgencyTactics": self.has_urgency_tactics,
            "hasSuspiciousLinks": self.has_suspicious_links,
            "hasImpersonation": self.has_impersonation,
            "requestsRemoteAccess": self.requests_remote_access,
            "requestsCredentialsOrOtp": self.requests_credentials_or_otp,
        }

def analyze_with_local_heuristics(message: str) -> Dict[str, Any]:
    """
    Authoritative local heuristic fallback engine.
    Scans for high-confidence Indian banking scam patterns, phishing vectors, and APK payloads.
    """
    text = message.lower()
    score = 0
    categories: List[str] = []
    reasons: List[str] = []
    actions: List[str] = []

    # Indicators
    has_urgency = False
    has_links = False
    has_impersonation = False
    requests_remote = False
    requests_credentials = False

    # 1. Impersonation detection
    impersonation_patterns = [
        "sbi", "hdfc", "icici", "axis", "punjab national", "rbi",
        "reserve bank", "police", "electricity office", "power department",
        "customer care", "bank manager", "kyc officer", "telecom"
    ]
    matched_org = [org for org in impersonation_patterns if org in text]
    if matched_org:
        score += 25
        has_impersonation = True
        categories.append("Institutional Impersonation")
        reasons.append(f"Impersonates authoritative bodies or utility providers ({', '.join(matched_org).upper()})")

    # 2. Urgency and Coercion
    urgency_patterns = [
        "blocked today", "suspended", "immediately", "within 24 hours",
        "power will be disconnected", "disconnected tonight", "urgent",
        "action required", "legal action", "arrest warrant", "freeze"
    ]
    matched_urgency = [u for u in urgency_patterns if u in text]
    if matched_urgency:
        score += 30
        has_urgency = True
        categories.append("Urgency & Coercion")
        reasons.append("Creates artificial panic with immediate forfeiture, arrest, or disconnection deadlines")

    # 3. Malicious Links / APK Payloads
    has_url = bool(re.search(r"https?://\S+|www\.\S+|bit\.ly/\S+|\.apk", text))
    if has_url or ".apk" in text or "download" in text and "link" in text:
        score += 35
        has_links = True
        categories.append("Phishing Link / Malicious APK")
        reasons.append("Directs user to click suspicious shortened link or download an untrusted APK package")

    # 4. Remote Access / Screen Sharing
    remote_keywords = ["anydesk", "teamviewer", "quicksupport", "rustdesk", "screen share", "9-digit code"]
    matched_remote = [r for r in remote_keywords if r in text]
    if matched_remote:
        score += 45
        requests_remote = True
        categories.append("Remote Access Trojan / Screen Control")
        reasons.append(f"Instructs victim to install remote desktop control software ({', '.join(matched_remote)})")

    # 5. Credential / OTP Harvest
    credential_keywords = ["otp", "pin", "pan card", "aadhaar", "password", "cvv", "card number", "debit card details"]
    matched_cred = [c for c in credential_keywords if c in text]
    if matched_cred:
        score += 35
        requests_credentials = True
        categories.append("Credential Harvesting")
        reasons.append(f"Demands sensitive banking identifiers or authentication factors ({', '.join(matched_cred)})")

    # Calculate final score and tier
    score = min(100, max(5, score))

    if score >= 75:
        risk_level = "CRITICAL" if score >= 85 else "HIGH"
        actions = [
            "DO NOT click any embedded links or install recommended applications.",
            "DO NOT share OTPs, PINs, or 9-digit remote control codes.",
            "Verify directly with your official bank branch or dial National Cybercrime Helpline 1930."
        ]
        recommended_action = "IMMEDIATELY DISREGARD AND BLOCK SENDER. Report incident to 1930 Cybercrime Helpline."
    elif score >= 40:
        risk_level = "WARN"
        actions = [
            "Treat with caution. Contact the sender through a verified telephone number.",
            "Do not disclose confidential bank account numbers or passwords."
        ]
        recommended_action = "EXERCISE CAUTION. Independently verify the message authenticity before taking action."
    else:
        risk_level = "SAFE"
        categories.append("Legitimate / Low Risk")
        reasons.append("No active phishing patterns, coercive threats, or credential harvest signatures identified.")
        actions = ["Standard banking precautions apply."]
        recommended_action = "Message appears informational. Always refrain from sharing UPI PINs to receive money."

    indicators = ThreatIndicators(
        has_urgency_tactics=has_urgency,
        has_suspicious_links=has_links,
        has_impersonation=has_impersonation,
        requests_remote_access=requests_remote,
        requests_credentials_or_otp=requests_credentials,
    )

    primary_category = categories[0] if categories else "General Security Alert"

    return {
        "score": score,
        "riskLevel": risk_level,
        "categories": categories,
        "reasons": reasons,
        "recommendedAction": recommended_action,
        "source": "LOCAL_HEURISTIC_ENGINE",
        "scamType": primary_category,
        "summary": f"Heuristic analysis flagged {len(categories)} security factors with {score}% threat probability.",
        "warningSigns": reasons,
        "actionAdvice": actions,
        "threatIndicators": indicators.to_dict(),
    }

async def analyze_scam_with_gemini(message: str) -> Dict[str, Any]:
    """
    Analyzes an inbound SMS, WhatsApp message, or call transcript using Gemini AI.
    Gracefully falls back to local heuristic analysis if API key is not configured or on network error.
    """
    api_key = settings.GEMINI_API_KEY or os.getenv("GEMINI_API_KEY", "")
    if not api_key:
        return analyze_with_local_heuristics(message)

    try:
        from google import genai
        from google.genai import types

        client = genai.Client(api_key=api_key)

        prompt = f"""
You are the Chief Financial Cybercrime Analyst for GuardianPay AI, an advanced digital banking safety layer.
Evaluate the following SMS, email, or WhatsApp message for scam and social engineering indicators:

MESSAGE CONTENT:
\"\"\"{message}\"\"\"

Analyze the message strictly and return ONLY a valid JSON object with the following structure:
{{
  "score": <integer from 0 to 100 representing scam risk>,
  "riskLevel": "<SAFE | WARN | HIGH | CRITICAL>",
  "scamType": "<e.g. Phishing SMS, Utility Bill Scam, KYC Impersonation, Remote Access Fraud, Legitimate>",
  "categories": ["<category1>", "<category2>"],
  "summary": "<one sentence clear assessment>",
  "reasons": ["<bullet 1>", "<bullet 2>"],
  "warningSigns": ["<sign 1>", "<sign 2>"],
  "actionAdvice": ["<action 1>", "<action 2>"],
  "recommendedAction": "<concise clear instruction for an elderly or non-tech-savvy user>",
  "threatIndicators": {{
    "hasUrgencyTactics": <true/false>,
    "hasSuspiciousLinks": <true/false>,
    "hasImpersonation": <true/false>,
    "requestsRemoteAccess": <true/false>,
    "requestsCredentialsOrOtp": <true/false>
  }}
}}
"""

        response = client.models.generate_content(
            model="gemini-2.5-flash",
            contents=prompt,
            config=types.GenerateContentConfig(
                response_mime_type="application/json",
            ),
        )

        response_text = response.text or ""
        # Clean up JSON markdown fences if present
        clean_json = re.sub(r"^```json\s*", "", response_text.strip())
        clean_json = re.sub(r"\s*```$", "", clean_json.strip())

        parsed = json.loads(clean_json)
        parsed["source"] = "GEMINI"
        return parsed

    except Exception as e:
        # On error, timeout, or quota limit, fall back safely to deterministic heuristic engine
        print(f"[GuardianPay AI] Gemini API call deferred to heuristic engine: {e}")
        fallback = analyze_with_local_heuristics(message)
        fallback["source"] = "LOCAL_HEURISTIC_ENGINE"
        return fallback

File: src/server/ai/geminiScamService.ts
import { GoogleGenAI } from '@google/genai';

export interface ScamAnalysisOutput {
  score: number;
  riskLevel: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  categories: string[];
  reasons: string[];
  recommendedAction: string;
  source: 'GEMINI_AI' | 'LOCAL_HEURISTIC_ENGINE';
}

export function analyzeScamLocalHeuristics(message: string): ScamAnalysisOutput {
  const lower = message.toLowerCase();
  let score = 5;
  const categories: string[] = [];
  const reasons: string[] = [];

  // Keywords checks
  const isOtpPin = /(otp|pin|cvv|password|passcode|verification code|security code)/i.test(lower);
  const isUrgent = /(immediately|urgent|today|tonight|within \d+|hours|minutes|expire|blocked|suspended|seized|disconnected)/i.test(lower);
  const isBank = /(sbi|hdfc|icici|axis|bank|rbi|yono|kyc|pan card|aadhaar|account)/i.test(lower);
  const isPower = /(electricity|power|bescom|mseb|tneb|light bill|meter)/i.test(lower);
  const isRemote = /(anydesk|teamviewer|quicksupport|screen share|install apk|\.apk|download app)/i.test(lower);
  const isPoliceLegal = /(police|cbi|customs|arrest|warrant|court|narcotics|digital arrest|fir)/i.test(lower);
  const isLotteryRefund = /(lottery|won|cashback|prize|reward|refund|credit of rs|claim)/i.test(lower);
  const hasUrl = /(https?:\/\/|bit\.ly|tinyurl|\.apk|\.xyz|\.top|\.online|[0-9]{10})/i.test(lower);

  if (isOtpPin) {
    score += 35;
    categories.push('Credential / OTP Harvesting');
    reasons.push('Requests sensitive one-time password (OTP), PIN, or security codes.');
  }

  if (isUrgent) {
    score += 25;
    categories.push('Artificial Urgency Coercion');
    reasons.push('Uses manufactured time pressure or immediate threats to bypass rational thought.');
  }

  if (isBank && (lower.includes('block') || lower.includes('kyc') || lower.includes('pan'))) {
    score += 30;
    categories.push('Fake Bank KYC / Account Block Scam');
    reasons.push('Impersonates financial institution threatening account suspension for pending KYC.');
  }

  if (isPower) {
    score += 30;
    categories.push('Utility Disconnection Fraud');
    reasons.push('Threatens immediate power disconnection to trick victim into calling unofficial number.');
  }

  if (isRemote) {
    score += 35;
    categories.push('Malicious APK / Remote Screen Access');
    reasons.push('Requests installation of APK or remote-control tools (AnyDesk/TeamViewer).');
  }

  if (isPoliceLegal) {
    score += 40;
    categories.push('Coercive Impersonation / Fake Legal Threat');
    reasons.push('Threatens law enforcement or arrest to panic senior citizens.');
  }

  if (isLotteryRefund) {
    score += 20;
    categories.push('Unsolicited Prize / Advance Refund Fraud');
    reasons.push('Promises unverified monetary lottery or refund requiring payment.');
  }

  if (hasUrl) {
    score += 15;
    categories.push('Deceptive Link / Unofficial Contact');
    reasons.push('Contains unofficial hyperlinks, APK downloads, or personal mobile numbers.');
  }

  score = Math.min(100, Math.max(5, score));

  let riskLevel: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL' = 'LOW';
  let recommendedAction = 'Message appears standard. Exercise normal caution.';

  if (score >= 80) {
    riskLevel = 'CRITICAL';
    recommendedAction =
      'DO NOT ACT. Do not click any links, do not install applications, and do not call back. Report immediately to National Cyber Helpline 1930.';
  } else if (score >= 60) {
    riskLevel = 'HIGH';
    recommendedAction =
      'HIGH RISK. Verify directly by visiting your official bank branch or dialing the official helpline listed on your bank card.';
  } else if (score >= 35) {
    riskLevel = 'MEDIUM';
    recommendedAction =
      'SUSPICIOUS. Cross-check this message with a trusted family member before taking any financial action.';
  }

  if (categories.length === 0) {
    categories.push('Routine Communication');
    reasons.push('No coercive patterns, deceptive links, or credential requests detected.');
  }

  return {
    score,
    riskLevel,
    categories,
    reasons,
    recommendedAction,
    source: 'LOCAL_HEURISTIC_ENGINE',
  };
}

export async function analyzeScamWithGemini(message: string): Promise<ScamAnalysisOutput> {
  const apiKey = process.env.GEMINI_API_KEY;

  if (!apiKey || apiKey.trim() === '') {
    console.log('[AI Engine] GEMINI_API_KEY not set. Using local deterministic heuristic engine.');
    return analyzeScamLocalHeuristics(message);
  }

  try {
    const ai = new GoogleGenAI({ apiKey });
    const prompt = `You are the GuardianPay AI Senior Fraud Specialist. Analyze the following SMS or message received by an Indian senior citizen banking customer.
Message:
"""
${message}
"""

Evaluate social engineering, psychological urgency, fake KYC, utility threats, and OTP/PIN harvesting.
Return ONLY a valid JSON object matching this exact TypeScript structure:
{
  "score": <number between 0 and 100>,
  "riskLevel": <"LOW" | "MEDIUM" | "HIGH" | "CRITICAL">,
  "categories": [<array of specific scam categories, e.g. "Fake Bank KYC", "Urgency Coercion">],
  "reasons": [<array of clear, plain-language explanation bullet points for an elderly victim>],
  "recommendedAction": <string of specific defensive action to take, e.g. "DO NOT ACT. Dial 1930.">
}
Do not include markdown code fence formatting or explanations outside JSON.`;

    const aiPromise = ai.models.generateContent({
      model: 'gemini-3.8-flash',
      contents: prompt,
      config: {
        responseMimeType: 'application/json',
      },
    });

    const timeoutPromise = new Promise<never>((_, reject) =>
      setTimeout(() => reject(new Error('Gemini API request timed out (4s)')), 4000)
    );

    const response = (await Promise.race([aiPromise, timeoutPromise])) as any;

    const text = response.text?.trim() || '';
    const cleaned = text.replace(/^```json/i, '').replace(/^```/, '').replace(/```$/, '').trim();
    const parsed = JSON.parse(cleaned);

    if (
      typeof parsed.score === 'number' &&
      parsed.riskLevel &&
      Array.isArray(parsed.categories) &&
      Array.isArray(parsed.reasons) &&
      typeof parsed.recommendedAction === 'string'
    ) {
      return {
        score: Math.min(100, Math.max(0, parsed.score)),
        riskLevel: parsed.riskLevel,
        categories: parsed.categories,
        reasons: parsed.reasons,
        recommendedAction: parsed.recommendedAction,
        source: 'GEMINI_AI',
      };
    }

    console.warn('[AI Engine] Gemini response did not match schema. Falling back to local engine.');
    return analyzeScamLocalHeuristics(message);
  } catch (err) {
    console.warn('[AI Engine] Gemini API request failed, seamlessly falling back to local heuristics:', err);
    return analyzeScamLocalHeuristics(message);
  }
}
